Fix form association behavior when a form and a control with form= are removed from a document together

rniwa opened the pull request:
#### d16d7ec796839625be5b5a686c5c82b55ba620b3
<pre>
Fix form association behavior when a form and a control with form= are removed from a document together
<a href="https://bugs.webkit.org/show_bug.cgi?id=247593">https://bugs.webkit.org/show_bug.cgi?id=247593</a>

Reviewed by Darin Adler.

Fix the bug that a form associated element with form content attribute were not disassociated from
its form owner if both elements are removed together in a same subtree.

New behavior matches the spec as well as behaviors of Chrome and Firefox:
<a href="https://html.spec.whatwg.org/multipage/infrastructure.html#html-element-removing-steps">https://html.spec.whatwg.org/multipage/infrastructure.html#html-element-removing-steps</a>
<a href="https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#reset-the-form-owner">https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#reset-the-form-owner</a>

* LayoutTests/fast/forms/form-associated-element-removal-with-form-attribute-expected.txt: Added.
* LayoutTests/fast/forms/form-associated-element-removal-with-form-attribute.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/form-control-form-attribute-expected.txt: Rebaselined.

* Source/WebCore/html/FormAssociatedElement.cpp:
(WebCore::FormAssociatedElement::removedFromAncestor):

Canonical link: <a href="https://commits.webkit.org/256620@main">https://commits.webkit.org/256620@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ca3d3a443946504be05ad307cdf96e2868baedbc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/5591 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29394 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105879 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166222 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100319 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5738 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34336 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88717 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/102609 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102008 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/4258 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82934 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31265 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86100 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88005 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/74123 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/40068 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/19508 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37744 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20884 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/1293 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/43462 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/2196 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/246 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40152 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/FormAssociatedElement.cpp:
```diff
@@ -89,15 +89,19 @@ void FormAssociatedElement::insertedIntoAncestor(Node::InsertionType insertionTy
-void FormAssociatedElement::removedFromAncestor(Node::RemovalType, ContainerNode&)
+void FormAssociatedElement::removedFromAncestor(Node::RemovalType removalType, ContainerNode&)
 {
+    HTMLElement& element = asHTMLElement();
```
`auto&`

rniwa replied:
Done.

darinadler commented on Source/WebCore/html/FormAssociatedElement.cpp:
```diff
@@ -89,15 +89,19 @@ void FormAssociatedElement::insertedIntoAncestor(Node::InsertionType insertionTy
     // Otherwise, null out our form and remove ourselves from the form's list of elements.
     // Do not rely on rootNode() because our IsInTreeScope is outdated.
-    if (m_form && &asHTMLElement().traverseToRootNode() != &m_form->traverseToRootNode())
+    if ((m_form && &asHTMLElement().traverseToRootNode() != &m_form->traverseToRootNode())
```
Should use `element` here instead of calling `asHTMLElement()` a second time.

rniwa replied:
Good point. Fixed.

rniwa committed:
Fix form association behavior when a form and a control with form= are removed from a document together
https://bugs.webkit.org/show_bug.cgi?id=247593

Reviewed by Darin Adler.

Fix the bug that a form associated element with form content attribute were not disassociated from
its form owner if both elements are removed together in a same subtree.

New behavior matches the spec as well as behaviors of Chrome and Firefox:
https://html.spec.whatwg.org/multipage/infrastructure.html#html-element-removing-steps
https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#reset-the-form-owner

* LayoutTests/fast/forms/form-associated-element-removal-with-form-attribute-expected.txt: Added.
* LayoutTests/fast/forms/form-associated-element-removal-with-form-attribute.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/shadow-dom/form-control-form-attribute-expected.txt: Rebaselined.

* Source/WebCore/html/FormAssociatedElement.cpp:
(WebCore::FormAssociatedElement::removedFromAncestor):

Canonical link: https://commits.webkit.org/256620@main
